mlir/lib/Conversion/TosaToLinalg/TosaToLinalg.cpp

```diff
@@ -404,7 +445,31 @@ static Value createLinalgBodyCalculationForElementwiseOp(
         loc, elementTy, rewriter.getFloatAttr(elementTy, minApf));
     auto max = rewriter.create<arith::ConstantOp>(
         loc, elementTy, rewriter.getFloatAttr(elementTy, maxApf));
-    return clampFloatHelper(loc, args[0], min, max, rewriter);
+    auto result = clampFloatHelper(loc, args[0], min, max, rewriter);
+
+    auto clampOp = llvm::cast<tosa::ClampOp>(op);
+    const auto nanMode = clampOp.getNanMode();
+    // In the case of "PROPAGATE" semantics no compare and selection is
+    // required.
+    if (nanMode == "PROPAGATE")
+      return result;
+
+    // In the case of "IGNORE" semantics materialize a comparison
+    // of the current operand to the reduction which will return true for a NaN
+    // argument and then selects between the initial reduction value and the
+    // calculated result based on whether the argument is NaN or not. In pseudo
+    // code:
+    //
+    // reduce<op>(x, init):
+    //   result = op(init, x)
+    //   return init if x == NaN else result
+
+    // Unordered comparison of NaN against itself will always return true.
+    Value isNaN = rewriter.create<arith::CmpFOp>(
```

Comment:
Can you refactor this in `materializeBinaryNanCheckIfRequired` ?

Reply:
Maybe I'm misunderstanding what you mean here but the checks are different in the case of clamp vs. a binary check. 

The binary check requires two compare instructions and two select instructions. The clamp operation requires one compare instruction and one select instruction, so the semantics of how the NaN is handled is slightly different. 

Maybe these could be merged into some more general logic but I think there is some value in being explicit here since it is easy to follow.

Reply:
I think you can pack a `CmpFOp` and `SelectOp` and reuse this with as many arguments.
Eitherway, looks fine; but would be great if you could see if this logic could be improved and patch this up.

Reply:
If we find a way to do this we can open a future PR to refactor the logic.